KAFKA-5903: Added Connect metrics to the worker and distributed herder

rhauch opened the pull request:
Added metrics to the Connect worker and rebalancing metrics to the distributed herder.

This is built on top of #3987, and I can rebase this PR once that is merged.

rhauch committed:
KAFKA-5903 Use string-valued metrics for connector name, type, version, and status

Changed from the indicator metrics for connector and task status that reported the various boolean states to a single string-valued metric for status. Also added string-valued metrics for connector name, version, and type.

rhauch committed:
KAFKA-5903 Added Connect worker metrics

rhauch committed:
KAFKA-5904 Added Connect rebalancing metrics to distributed herder

Added the rebalancing metrics to the distributed herder.

rhauch commented:
This is the final PR for the Connect metrics.

The rebalance metrics don't really make sense on the standalone herder, so I only added them to the distributed herder. The KIP can be changed to reflect this. 

Also, the KIP included worker metrics for task startup counts (successes, failures, total, etc.) but neglected to define metrics for connector startup counts. This PR adds them.

I can update the KIP to reflect all of these changes once this is merged.

rhauch commented:
@ewencp, please have a look at this PR for the last Connect metrics. Thanks!!

rhauch commented:
@ewencp rebased on `trunk` and updated this PR to fix the merge conflicts.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java:
```diff
@@ -230,13 +246,57 @@ public ConnectMetricsRegistry(Set<String> tags) {
+
+        connectorCount = createTemplate("connector-count", WORKER_GROUP_NAME, "The number of connectors run in this worker.", workerTags);
+        taskCount = createTemplate("task-count", WORKER_GROUP_NAME, "The number of tasks run in this worker.", workerTags);
+        workerStatus = createTemplate("status", WORKER_GROUP_NAME, "The state of this worker. One of: 'running', 'stopping', or 'stopped'.",
```
Hmm, this seems different than the KIP and seems to provide less information, e.g. rebalancing never shows up. I figured this would either reuse the enum from AbstractCoordinator or put a new one in AbstractHerder that the herders then update. Despite the class name matching, I think this should really refer to the worker process rather than the state of the `Worker` class. The class is really just a container for connectors/tasks so it doesn't have enough information to understand the state of the whole process.

rhauch replied:
The problem is that the worker doesn't really known when a rebalance is occurring -- the herders uses the worker, but the distributed worker doesn't really inform the worker when it is rebalancing. That's why I added the `rebalancingMode` to the rebalance metrics, which returns 1 when rebalancing or 0 when not. WDYT?

ewencp replied:
ok, so if the info is available somewhere else that makes sense. then does this metric really add any value? it's only not `running` very early after the process starts or briefly when shutting down. I guess it kind of helps debug if you get stuck doing something at startup/shutdown, but is that useful as a metric given that you just have to look at logs anyway?

rhauch replied:
Good question about whether it's valuable. About the only thing it might help identify is a worker process that has been told to stop but is blocked waiting for any connectors and/or tasks to stop (in [lines 166-172](https://github.com/apache/kafka/pull/4011/files#diff-316d2c222b623ee65e8065863bf4b9ceR166-172)).

rhauch replied:
@ewencp okay, since the worker's `status` metric has virtually no value, it's better to not have it. Added a commit that removes it.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java:
```diff
@@ -103,17 +106,13 @@ public void setup() {
         workerProps.put("internal.key.converter.schemas.enable", "false");
         workerProps.put("internal.value.converter.schemas.enable", "false");
         workerProps.put("offset.storage.file.filename", "/tmp/connect.offsets");
+        workerProps.put("offset.storage.file.filename", "/tmp/connect.offsets");
```
seems this line got duplicated?

rhauch replied:
ack. will remove.

ewencp reviewed (commented):
@rhauch pretty close, just one question re: difference in metric from the KIP

rhauch committed:
KAFKA-5903 Removed duplicate config line

rhauch committed:
KAFKA-5903 Removed Connect worker’s `state` metric

The metric adds almost no value, since it is always `RUNNING` except in the few moments just before the worker process stops.

ewencp reviewed (approved):
Great, LGTM now. Will just wait on tests to come back green.
